### PR TITLE
Introduce generic reconciliation framework

### PR DESCRIPTION
**Pull request by anthony-gomez-fastly**
### Change summary

Introduce internal/reconcile package with generic Resource[T, API] to eliminate
  duplicated list→diff→create/update/delete patterns across backend and domain resources.

  Changes:
  - Add internal/reconcile package with generic Resource[T, API] reconciler
    * Run() method for reconciliation (list, diff, create/update/delete)
    * ReadForVersion() method for reading and converting API objects
    * ModelsEqual() helper for comparing slices with optional sorting
    * Comprehensive test suite covering create, update, delete, edge cases

  - Refactor backend resource to use reconciler
    * Move reconcile adapter ops into schema.go (consolidate files)
    * Reduce ReadForVersion from 17 lines to 1 line (delegates to reconciler)
    * Reduce Reconcile from 59 lines to 1 line (delegates to reconciler)
    * Reduce Equal from 13 lines to 1 line (uses reconcile.ModelsEqual)
    * Remove Normalize function (handled internally by reconciler)
    * Remove TestNormalize (behavior now tested via reconciler)
    * Replace map copy loop with maps.Copy

  - Refactor domain resource to use reconciler
    * Move reconcile adapter ops into schema.go (consolidate files)
    * Reduce ReadForVersion from 17 lines to 1 line (delegates to reconciler)
    * Reduce Reconcile from 79 lines to 1 line (delegates to reconciler)
    * Reduce Equal from 28 lines to 1 line + 16-line helper (uses reconcile.ModelsEqual)
    * Remove Normalize function (handled internally by reconciler)
    * Remove TestNormalize (behavior now tested via reconciler)
    * Replace map copy loop with maps.Copy

 

 All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/fastly/terraform-provider-fastly/pulls) for the same update/change?

### New Feature Submissions:

* [x] Does your submission pass tests?
* [x] Post the output of your test runs
Screenshot 2026-06-23 at 11 21 48 AM

### CODE REVIEW

File: internal/acceptance_tests/config_builder.go [hunk 1]
```diff
@@ -67,9 +68,7 @@ func BuildConfig(serviceType ServiceType, replacements map[string]string, blockP
 
 	// Build template data with all replacements
 	templateData := make(map[string]string)
-	for k, v := range replacements {
-		templateData[k] = v
-	}
+	maps.Copy(templateData, replacements)
```
Comment (anthony-gomez-fastly) [inline internal/acceptance_tests/config_builder.go:71]: this was added in 16.4, decided to update since i was touching files with the same nit

File: internal/reconcile/reconcile.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+func (r *Resource[T, API]) Run(ctx context.Context, client *fastly.Client, serviceID string, version int, desired []T) error {
+	remote, err := r.Ops.List(ctx, client, serviceID, version)
+	if err != nil {
+		return err
+	}
+
+	if r.Sortable {
+		sort.Slice(desired, func(i, j int) bool {
```
Comment (philippschulte) [inline internal/reconcile/reconcile.go]: This can mutate the caller’s slice unexpectedly. Since this reconciler is generic and may be reused in more places later, it would be safer to copy before sorting

**Review by philippschulte — DISMISSED** [review]
Nice refactor.

One small request before merge: please copy the slice before sorting in `Resource.Run()`. Sorting the `desired` slice in place can mutate the caller’s backing array, which could lead to unexpected ordering changes outside the reconciler.

Once this is addressed, this is good to merge from my side.

Thanks!

File: internal/reconcile/reconcile.go (hunk 2 above)
Comment (anthony-gomez-fastly) [reply internal/reconcile/reconcile.go]: fixed in `0c079919`

**Review by philippschulte — APPROVED** [review]
Thank you!

File: internal/resources/domain/schema.go [hunk 3]
```diff
@@ -59,153 +57,125 @@ func NestedBlockSchema() schema.ListNestedBlock {
...
+	name := desired.Name.ValueString()
+	input := &fastly.CreateDomainInput{
 		ServiceID:      serviceID,
 		ServiceVersion: version,
-	})
-	if err != nil {
-		return err
+		Name:           &name,
+	}
+	if !desired.Comment.IsUnknown() {
```
Comment (rcaril) [inline internal/resources/domain/schema.go]: Should we also add a `IsNull` guard here like we have on line `83`?

File: internal/reconcile/reconcile.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]T, 0, len(remote))
+	for _, item := range remote {
+		result = append(result, r.Ops.ToModel(item))
+	}
+
+	if r.Sortable {
```
Comment (rcaril) [inline internal/reconcile/reconcile.go:97]: Sortable conflicts with CDTOOL-1665:

> The Sortable: true flag causes ReadForVersion to sort results before writing them to Terraform state. This is the same issue raised in CDTOOL-1665 -  `ListNestedBlock` is positional, so if a user configures `backend[0] = b`, `backend[1] = a`, the provider cannot reorder and return `[a, b] `in state without Terraform throwing "Provider produced inconsistent result after apply." The sort in Run is also a no-op - the reconcile loop is entirely map-based (remoteByName/desiredByName), so iteration order of desired never affects which items get created, updated, or deleted. Sortable belongs only in `ModelsEqual` where you control both sides for internal comparison.

We should consider holding off on this PR until we update the sorting issue first.

File: internal/reconcile/reconcile_test.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	}
+	return o.listResult, nil
+}
+
+func (o *testOps) GetName(api *testAPI) string {
+	return api.Name
+}
+
+func (o *testOps) Delete(ctx context.Context, client *fastly.Client, serviceID string, version int, name string) error {
+	if o.deleteError != nil {
```
Comment (rcaril) [inline internal/reconcile/reconcile_test.go:45]: Should we add tests for the `deleteError`, `createError` and `updateError` methods? I believe we already do this elsewhere such as on line `170`:

```
t.Run("handles list error", func(t *testing.T) {
		ops := &testOps{
			listError: errors.New("list failed"),
		}
```
                                                                                                                               
 Perhaps something like:                                                                                                                           
  ```                                                                                                                          
  t.Run("returns error when delete fails", func(t *testing.T) {                                                                
        ops := &testOps{                                                                                                       
            listResult:  []*testAPI{{Name: "item1"}},                                                                          
            deleteError: errors.New("delete failed"),                                                                          
        }                                                                                                                      
        // desired is empty so delete is triggered                                                                             
        err := r.Run(ctx, nil, "service123", 1, []testModel{})                                                                 
        assert.Error(t, err)                                                                                                   
    })                                                                                                                         
  ```

**Review by rcaril — DISMISSED** [review]
Great work! A few items that may need to be reviewed below.

File: internal/reconcile/reconcile.go (hunk 4 above)
Comment (philippschulte) [reply internal/reconcile/reconcile.go:97]: @rcaril, @anthony-gomez-fastly I'll address CDTOOL-1665 as soon as this pr is merged. I am just waiting for this pull request to be merged. Thanks

File: internal/reconcile/reconcile_test.go (hunk 5 above)
Comment (anthony-gomez-fastly) [reply internal/reconcile/reconcile_test.go:45]: addressed in `fcc189f2`

File: internal/resources/domain/schema.go (hunk 3 above)
Comment (anthony-gomez-fastly) [reply internal/resources/domain/schema.go]: addressed in `a09639db`
